### PR TITLE
(GH-253) Package Test Results / Package Verifier interaction

### LINKED ISSUE #214: Package review process should take place on the site
We all have certain passions, and certain things we really _really_ couldn't care less about. I'm hypothesizing that although we all have _different_ packages that we are passionate about, we _all_ have a huge amount of packages that we would personally flag as either "irrelevant" or even "crap".

When a moderator is - temporary or permanently - having a busy life, reading the vast stream of mails to filter out the packages you personally find most valuable and willing to verify, or searching for the ones where you are personally mentioned or addressed as a moderator, becomes too tedious of a job and the moderator becomes tempted to swipe all (AKA ignore all or mark as read).

I propose that one should have the option to mark certain packages and only receive correspondence about them. This preference should also traverse to the moderation queue so that you only see packages relevant to your interests in the moderation queue.

This is probably not benefiting new, less known, and irrelevant packages. But at least it would give attention to some packages, and as it is now, we're still dealing with a huge backlog, even on more popular packages.

### LINKED ISSUE #253: Support for package verifier service
Similar to #181 and #175 but introduces the changes for the package verifier.

### PR DESCRIPTION
**Pull request by ferventcoder**
Closes #253
Closes #214

### CODE REVIEW

File: chocolatey/Website/Entities/PackageTestResults.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.ComponentModel.DataAnnotations;
+
+namespace NuGetGallery
+{
+    //note: don't store this in the database as a separate entity yet.
```
Comment (gep13) [inline chocolatey/Website/Entities/PackageTestResults.cs]: @ferventcoder I am assuming that this is going to get uncommented at some point?  Is it required in this PR?

**Comment by gep13** [comment]
@ferventcoder looks good to me, only one comment/question.

File: chocolatey/Website/Entities/PackageTestResults.cs (hunk 1 above)
Comment (ferventcoder) [inline chocolatey/Website/Entities/PackageTestResults.cs]: At some point. While we are only testing on one system I didn't see a need to introduce it quite yet.

**Comment by ferventcoder** [comment]
@gep13 would prefer less glossy icons for the balls.

Unknown:
image

Pending:
image

Failed:
image

Passed:
image

**Comment by ferventcoder** [comment]
On failed and passed, the ball doubles as a link.

File: chocolatey/Website/Entities/PackageTestResults.cs (hunk 1 above)
Comment (gep13) [inline chocolatey/Website/Entities/PackageTestResults.cs]: Makes sense.

**Comment by gep13** [comment]
@ferventcoder I actually kind of like the icons as they are :smile:

**Comment by ferventcoder** [comment]
@gep13 sweet. sold then.

**Comment by ferventcoder** [comment]
Maintainer response 

image


Maintainer attempting to reject without comments:
image

**Comment by gep13** [comment]
Couple comments:

Maintainer is able to click the Submit button, even though no comments have been entered.  When this happened, message is reported back to the page to indicate that submission was successful.

Once a package has been rejected by the maintainer, the "Reject Package" option is still available for selection.  While in the rejected state, this option should not be there.

**Comment by gep13** [comment]
Should I be able to edit this section:

image

I would have thought a moderator would only be able to add additional comments, rather than edit old ones?  Or are old comment updates sent through to the maintainers as well?

**Comment by gep13** [comment]
In the second action/comment in the screenshot above, as a maintainer, I changed the status from Submitted to Rejected.  Should/can this status change be logged in the review comments as well?

**Comment by ferventcoder** [comment]
@gep13 only an admin can edit the original review comments. 

Make sure you have a maintainer, moderator, and an admin account to look through with.

**Comment by gep13** [comment]
Can we get a non-ambiguous DateTime Format string for the review comments?  I personally find what is currently being used really hard to read.  Just my two cents...

**Comment by gep13** [comment]
>@ferventcoder said...
Make sure you have a maintainer, moderator, and an admin account to look through with.

ah, gotcha!

**Comment by ferventcoder** [comment]
> Can we get a non-ambiguous DateTime Format string for the review comments? I personally find what is currently being used really hard to read. Just my two cents...

Please tell me your preference. The current one is good for any locale, it just might be hard to read.

**Comment by ferventcoder** [comment]
> I would have thought a moderator would only be able to add additional comments, rather than edit old ones? Or are old comment updates sent through to the maintainers as well?

Old comment updates are still sent to maintainers as well.

**Comment by ferventcoder** [comment]
> Maintainer is able to click the Submit button, even though no comments have been entered. When this happened, message is reported back to the page to indicate that submission was successful.

> Once a package has been rejected by the maintainer, the "Reject Package" option is still available for selection. While in the rejected state, this option should not be there.

Thanks!

**Comment by ferventcoder** [comment]
> In the second action/comment in the screenshot above, as a maintainer, I changed the status from Submitted to Rejected. Should/can this status change be logged in the review comments as well?

I'm guessing it isn't. I think we should have all rejections require comments.

**Comment by gep13** [comment]
I personally prefer something like this:

image

**Comment by ferventcoder** [comment]
That works as well.

**Comment by gep13** [comment]
>@ferventcoder said...
I'm guessing it isn't. I think we should have all rejections require comments.

Agreed, but in addition to the comment that the user provides, can Chocolatey.org add something like:

`Status Changed to Rejected, with following comment: <insert comment from user here>`

Same when a moderator/admin changes the state of a package version.

**Comment by ferventcoder** [comment]
You want status change in there for every time a status changes?

**Comment by gep13** [comment]
From a tractability stand point, to see what has happened, when and why, I think it could be useful yes.

Related to that though...

Are you only showing the top x review comments, or is this being truncated?

**Comment by gep13** [comment]
Ok, I now see what you mean about the difference between each user, i.e. moderator, admin and reviewer :+1:

**Comment by gep13** [comment]
When logged in as a reviewer, I can see this:

image

Not sure if this is required here, as no action can be taken once expanded.

**Comment by ferventcoder** [comment]
> Are you only showing the top x review comments, or is this being truncated?

All comments

**Comment by gep13** [comment]
@ferventcoder I am not set up to test the actual package interaction, need you to help me set that up locally at some point so I can have a play with it, but aside from the comments above, this looks sound to me!  REALLY, REALLY like this addition, great work!

**Comment by ferventcoder** [comment]
> I am not set up to test the actual package interaction, need you to help me set that up locally at some point

You mean like emails and bits? You can just push to the local api with packages and they are saved on the file path under app_data. So are emails from the system (saved as files).

**Comment by gep13** [comment]
>@ferventcoder said...
All comments

This might become quite cumbersome for lengthy package moderation, but for a first release, I think this is fine :+1:

**Comment by gep13** [comment]
>@ferventcoder said...
You mean like emails and bits? You can just push to the local api with packages and they are saved on the file path under app_data. So are emails from the system (saved as files).

No, I meant with a local package verifier, for testing out all the globes appearing, and automatic installation comments, links to gists, etc.

**Comment by ferventcoder** [comment]
> No, I meant with a local package verifier, for testing out all the globes appearing, and automatic installation comments, links to gists, etc.

I don't have a system locally that can actually do this (my Windows laptops are too old to support Intel VT-x, which is required for x64 VMs).

**Comment by ferventcoder** [comment]
@gep13 I think I caught everything we discussed. Please take a look. :)

**Comment by ferventcoder** [comment]
Looks like I killed the link in older versions

**Comment by gep13** [comment]
Woot! :smile:

**Pull request merged**